File: HackerRank Solutions/Interview Preparation Kit/String Manipulation/Sherlock and the Valid String.py
#!/bin/python3

import math
import os
import random
import re
import sys
from collections import Counter

# Complete the isValid function below.
def isValid(s):
    characterFrequency = Counter(s)
    print(characterFrequency)
    frequencyOfCharacterFrequency = Counter(characterFrequency.values())
    print(frequencyOfCharacterFrequency)
    if len(frequencyOfCharacterFrequency) == 1 or len(frequencyOfCharacterFrequency) == 0:
        return 'YES'
    elif len(frequencyOfCharacterFrequency) > 2:
        return 'NO'
    else:
        if 1 in frequencyOfCharacterFrequency.values():
            if frequencyOfCharacterFrequency[min(frequencyOfCharacterFrequency.keys())] == 1 or (max(frequencyOfCharacterFrequency.keys()) - min(frequencyOfCharacterFrequency.keys()) == 1):
                return 'YES'
            else:
                return 'NO'
        else:
            return 'NO'

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    s = input()

    result = isValid(s)

    fptr.write(result + '\n')

    fptr.close()

File: HackerRank Solutions/Interview Preparation Kit/String Manipulation/Sherlock and the Valid String.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    // Complete the isValid function below.
    static String isValid(String s) {
        Map<Character, Integer> characterFrequency = new HashMap<Character, Integer>();
        for (char c : s.toCharArray()) {
            characterFrequency.put(c, characterFrequency.getOrDefault(c, 0) + 1);
        }
        Map<Integer, Integer> frequencyOfCharacterFrequency = new HashMap<Integer, Integer>();
        for (int i : characterFrequency.values()) {
            frequencyOfCharacterFrequency.put(i, frequencyOfCharacterFrequency.getOrDefault(i, 0) + 1);
        }
        if ((frequencyOfCharacterFrequency.size() == 0) || (frequencyOfCharacterFrequency.size() == 1)) {
            return "YES";
        } else if (frequencyOfCharacterFrequency.size() > 2) {
            return "NO";
        } else {
            if (frequencyOfCharacterFrequency.values().contains(1)) {
                if (frequencyOfCharacterFrequency.get(Collections.min(frequencyOfCharacterFrequency.keySet())) == 1 || (Collections.max(frequencyOfCharacterFrequency.keySet()) - Collections.min(frequencyOfCharacterFrequency.keySet()) == 1)) {
                    return "YES";
                } else {
                    return "NO";
                }
            } else {
                return "NO";
            }
        }
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        String s = scanner.nextLine();

        String result = isValid(s);

        bufferedWriter.write(result);
        bufferedWriter.newLine();

        bufferedWriter.close();

        scanner.close();
    }
}
